src/envoy/prototype/api_manager_env.cc

```diff
@@ -129,20 +127,82 @@ class RequestCallbacks : public AsyncClient::Callbacks {
   }
 };
 
+namespace {
+// Copy the code here from envoy/grpc/common.cc
+Buffer::InstancePtr SerializeGrpcBody(const std::string &body_str) {
+  // http://www.grpc.io/docs/guides/wire.html
+  Buffer::InstancePtr body(new Buffer::OwnedImpl());
+  uint8_t compressed = 0;
+  body->add(&compressed, sizeof(compressed));
+  uint32_t size = htonl(body_str.size());
+  body->add(&size, sizeof(size));
+  body->add(body_str);
+  return body;
+}
+Http::MessagePtr PrepareGrpcHeaders(const std::string &upstream_cluster,
+                                    const std::string &service_full_name,
+                                    const std::string &method_name) {
+  Http::MessagePtr message(new Http::RequestMessageImpl());
+  message->headers().insertMethod().value(
+      Http::Headers::get().MethodValues.Post);
+  message->headers().insertPath().value(
+      fmt::format("/{}/{}", service_full_name, method_name));
+  message->headers().insertHost().value(upstream_cluster);
+  message->headers().insertContentType().value(Grpc::Common::GRPC_CONTENT_TYPE);
+  return message;
```

Comment:
You may have missed this line

 message->headers().addViaMoveValue(LowerCaseString("te"), "trailers");

according to @lizan , it will not work if you miss it.